File: 3739.count-subarrays-with-majority-element-ii.java
#
# @lc app=leetcode id=3739 lang=java
#
# [3739] Count Subarrays With Majority Element II
#

# @lc code=start
class Solution {
    public long countMajoritySubarrays(int[] nums, int target) {
        int n = nums.length;
        int[] score = new int[n + 1];
        // Step 1: Build prefix scores: +1 for target, -1 for others
        for (int i = 0; i < n; ++i) {
            score[i + 1] = score[i] + (nums[i] == target ? 1 : -1);
        }
        // Step 2: Coordinate compression
        int[] sorted = score.clone();
        java.util.Arrays.sort(sorted);
        java.util.Map<Integer, Integer> compress = new java.util.HashMap<>();
        int idx = 0;
        for (int val : sorted) {
            if (!compress.containsKey(val)) {
                compress.put(val, idx++);
            }
        }
        // Step 3: Fenwick Tree (Binary Indexed Tree)
        Fenwick fenwick = new Fenwick(idx);
        long ans = 0;
        for (int i = 0; i <= n; ++i) {
            int c = compress.get(score[i]);
            // Query: sum of counts of prefix scores less than current
            ans += fenwick.query(c - 1);
            // Update: add current prefix score
            fenwick.update(c, 1);
        }
        return ans;
    }
    // Fenwick Tree implementation
    static class Fenwick {
        long[] tree;
        int n;
        Fenwick(int n) {
            this.n = n + 2;
            tree = new long[this.n];
        }
        void update(int i, long delta) {
            for (++i; i < n; i += i & -i) tree[i] += delta;
        }
        long query(int i) {
            long res = 0;
            for (++i; i > 0; i -= i & -i) res += tree[i];
            return res;
        }
    }
}
# @lc code=end

File: 3739.count-subarrays-with-majority-element-ii.go
#
# @lc app=leetcode id=3739 lang=golang
#
# [3739] Count Subarrays With Majority Element II
#
# @lc code=start
func countMajoritySubarrays(nums []int, target int) int64 {
    n := len(nums)
    prefix := make([]int, n+1)
    for i, v := range nums {
        if v == target {
            prefix[i+1] = prefix[i] + 1
        } else {
            prefix[i+1] = prefix[i] - 1
        }
    }
    // To count for each i, number of j < i such that prefix[i] > prefix[j]
    // Use a map to count frequency of prefix sums
    // For efficient counting, use a sorted list of prefix sums
    // But since values can be negative, compress values
    allPrefix := make([]int, n+1)
    copy(allPrefix, prefix)
    // Coordinate compression
    valMap := make(map[int]int)
    uniq := make([]int, 0)
    for _, v := range allPrefix {
        valMap[v] = 1
    }
    for k := range valMap {
        uniq = append(uniq, k)
    }
    sort.Ints(uniq)
    for i, v := range uniq {
        valMap[v] = i
    }
    // Fenwick Tree
    type BIT struct {
        tree []int
    }
    func (b *BIT) add(i, x int) {
        for i < len(b.tree) {
            b.tree[i] += x
            i += i & -i
        }
    }
    func (b *BIT) sum(i int) int {
        res := 0
        for i > 0 {
            res += b.tree[i]
            i -= i & -i
        }
        return res
    }
    size := len(uniq) + 2
    bit := &BIT{tree: make([]int, size)}
    var ans int64 = 0
    for i := 0; i <= n; i++ {
        idx := valMap[prefix[i]] + 1
        // For i, number of prefix[j] < prefix[i] is sum(idx-1)
        ans += int64(bit.sum(idx - 1))
        bit.add(idx, 1)
    }
    return ans
}
# @lc code=end